### ISSUE TITLE
Invitation notification

### ISSUE THREAD

**Issue by z33ky** [Nheko-Reborn/nheko#277]
I do not notice when people invite me to rooms, since the application does not set the alert-Flag.
From looking at the code I believe there is also no desktop-notification for that event.

If my assessment is incorrect and there is a desktop-notification for it, adding a call to `QApplication::alert()` is simple enough.

Otherwise, it seems to me like receiving chat messages ("syncing") is handled separately from notifications by the Matrix protocol. I'm guessing to not send notifications to multiple devices accessing the same account?
Anyways, it seems notifications are designed around having `eventId`s, which come from a `mtx::responses::Notification`.
Invitations are a result of a `mtx::responses::Sync`, which do not have `eventId`s. `eventId`s are converted to strings for internal usage (for `Cache` and `NotificationManager`), so we can maybe craft our own out of the `roomId` and some identifier to prevent collisions with the `eventId`s for chat-messages (which I assume are contained in the `mtx::responses::Notification`s).
Alternatively functions (and Cache-tables) can be added to handle chat-message-notifications and invitation-notifications separately. The existing names for the Notifications (`pushNotification()`, `markSentNotification()`, `"sent_notifications"` (Cache table)) sound confusing if we add our own "non-Matrix" notifications. They should then maybe be reamed (e.g. `pushMtxNotification()`, `markSentMtxNotification()`, `"sent_mtx_notifications"` (of course `"sent_notifications"` needs to be renamed on application-startup for backwards compat)).

I'd probably be up for implementing this feature.

**Comment by deepbluev7** [comment]
Yes, there are no notifications for invites. It should be fairly easy to implement and I don't think you need to do as many major changes as you suggest:

You can still reuse the postNotification function. You can just leave the eventId empty and make up your own body, everything else should probably be the same. The event id is only used to deduplicate notifications, i.e. notify for every notification exactly once. Invites should only notify once, so using an empty event id or a special event id of "invite_<roomid>" should work too.

So the postNotification stuff is fairly simple, what you may consider to change though is all the other notification functions to take an eventId and roomid, so that you can actually pass an empty event id. And then you just add that to the sync response handler, when iterating the invite rooms. I don't think we have to rename anything. It's a notification, it barely matters, where it is from imo. And then when at the same location you can do the same altert logic as well. You may want to consider factoring that logic out into a function, so that you don't need to duplicate all of that.

If you decide to work on that, please tell me in advance, so we don't collide anything and consider asking questions in our #nheko room, if you have any or are stuck.

**Comment by z33ky** [comment]
Yeah, sounds good. I'll start working on this in the next few days.
I'll head over to #nheko if any questions arise, thanks.

**Comment by 3v1n0** [comment]
Any news here?

I always miss invitations 😄
